tunnel/intra/retrier.go

```diff
@@ -275,3 +288,52 @@ func (r *retrier) CloseWrite() error {
 	defer r.mutex.Unlock()
 	return r.conn.CloseWrite()
 }
+
+func (r *retrier) Close() error {
+	if err := r.CloseWrite(); err != nil {
+		return err
+	}
+	return r.CloseRead()
+}
+
+// LocalAddr behaves slightly strangely: its value may change as a
+// result of a retry.  However, LocalAddr is largely useless for
+// TCP client sockets anyway, so nothing should be relying on this.
+func (r *retrier) LocalAddr() net.Addr {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	return r.conn.LocalAddr()
+}
+
+func (r *retrier) RemoteAddr() net.Addr {
+	return r.addr
+}
+
+func (r *retrier) SetReadDeadline(t time.Time) error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	r.readDeadline = t
+	// Don't enforce read deadlines until after the retry
+	// is complete.  Retry relies on setting its own read
+	// deadline, and we don't want this to interfere.
+	if r.retryCompleted() {
+		return r.conn.SetReadDeadline(t)
+	}
+	return nil
+}
+
+func (r *retrier) SetWriteDeadline(t time.Time) error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	r.writeDeadline = t
+	return r.conn.SetWriteDeadline(t)
+}
+
+func (r *retrier) SetDeadline(t time.Time) error {
+	e1 := r.SetReadDeadline(t)
```

Comment:
Per the comment above, do you need to check whether `r.retryCompleted()` before setting the read deadline?